tools/block-generator/generator/generate.go

```diff
@@ -595,71 +494,69 @@ func getAppTxOptions() []interface{} {
 
 // ---- Transaction Generation (Pay/Asset/Apps) ----
 
-func (g *generator) generateSignedTxns(round uint64, intra uint64) ([]txn.SignedTxn, []txn.ApplyData, uint64 /* nextIntra */, error) {
-	// TODO: return the number of transactions generated instead of updating intra!!!
+func (g *generator) generateTxGroup(round uint64, intra uint64) ([]txn.SignedTxnWithAD, uint64 /* numTxns */, error) {
 	selection, err := weightedSelection(g.transactionWeights, getTransactionOptions(), paymentTx)
 	if err != nil {
-		return nil, nil, intra, err
+		return nil, 0, err
 	}
 
 	var signedTxns []txn.SignedTxn
-	var ads []txn.ApplyData
-	var nextIntra uint64
+	var numTxns uint64
 	var expectedID uint64
 	switch selection {
 	case paymentTx:
 		var signedTxn txn.SignedTxn
-		var ad txn.ApplyData
-		signedTxn, ad, nextIntra, err = g.generatePaymentTxn(round, intra)
+		signedTxn, numTxns, err = g.generatePaymentTxn(round, intra)
 		signedTxns = []txn.SignedTxn{signedTxn}
-		ads = []txn.ApplyData{ad}
 	case assetTx:
 		var signedTxn txn.SignedTxn
-		var ad txn.ApplyData
-		signedTxn, ad, nextIntra, expectedID, err = g.generateAssetTxn(round, intra)
+		signedTxn, numTxns, expectedID, err = g.generateAssetTxn(round, intra)
 		signedTxns = []txn.SignedTxn{signedTxn}
-		ads = []txn.ApplyData{ad}
 	case applicationTx:
-		signedTxns, ads, nextIntra, expectedID, err = g.generateAppTxn(round, intra)
+		signedTxns, numTxns, expectedID, err = g.generateAppTxn(round, intra)
 	default:
-		return nil, nil, intra, fmt.Errorf("no generator available for %s", selection)
+		return nil, 0, fmt.Errorf("no generator available for %s", selection)
 	}
 
 	if err != nil {
-		return nil, nil, intra, fmt.Errorf("error generating transaction: %w", err)
+		return nil, numTxns, fmt.Errorf("error generating transaction: %w", err)
 	}
 
 	if len(signedTxns) == 0 {
-		return nil, nil, intra, fmt.Errorf("no transactions generated")
+		return nil, numTxns, fmt.Errorf("no transactions generated")
```

Comment:
maybe combine this with the err!=nil block?

Reply:
Now that we have `errors.Join()` it's easier to do that.

Reply:
But as I was implementing joining the errors, I realized that the first error condition (coming from `g.generateAppTxn()`) would result in always reporting that the number of `signedTxns` as zero, so wouldn't be very informative. So I prefer to keep these separate but I've added a "this should never happen!" to the message:

c5ae50b174168c58678a7b5ac3fc2131c73e2c3c